src/storage/pooled_storage_manager.h

```diff
@@ -148,9 +157,21 @@ void GPUPooledStorageManager::Alloc(Storage::Handle* handle) {
     mxnet::common::cuda::DeviceStore device_store(handle->ctx.real_dev_id(), true);
     size_t free, total;
     cudaMemGetInfo(&free, &total);
-    if (free <= total * reserve_ / 100 || size > free - total * reserve_ / 100)
+    double mem_limit_in_bytes = total * memory_limit_percentage_ / 100.0;
+    free = mem_limit_in_bytes - used_memory_;
+    if (free <= mem_limit_in_bytes * reserve_ / 100 ||
+        size > free - mem_limit_in_bytes * reserve_ / 100)
       ReleaseAll();
 
+    if (used_memory_ + size > mem_limit_in_bytes) {
+      // This calls abort() unless
+      // DMLC_LOG_FATAL_THROW != 0, then it
+      // throws std::runtime_error()
+      LOG(FATAL) << "memory limit reached, used: " << used_memory_ << "  limit: "
+                 << mem_limit_in_bytes << " (" << memory_limit_percentage_ << "% of "
+                 << total << ")";
+    }
```

Comment:
Why?

Reply:
This is the enforcement of the user-specified GPU memory limit. If users say "MXNet can use 1GB", then it needs to someone let users know that it cant function properly anymore within the set memory limit.

Its not any different than running on a GPU with small amount of GPU memory than needed (artificial limit).